src/lib/scheduler/worker.cpp

```diff
@@ -62,46 +62,46 @@ void Worker::operator()() {
 
   _set_affinity();
 
-  while (Hyrise::get().scheduler()->active()) {
-    _work();
+  while (Hyrise::get().scheduler()->active() && _active) {
+    // Worker is allowed to sleep (when queue is empty) as long as the scheduler is not shutting down.
+    _work(AllowSleep::Yes);
   }
 }
 
-void Worker::_work() {
-  // If execute_next has been called, run that task first, otherwise try to retrieve a task from the queue.
+void Worker::_work(const AllowSleep allow_sleep) {
   auto task = std::shared_ptr<AbstractTask>{};
   if (_next_task) {
     task = std::move(_next_task);
     _next_task = nullptr;
   } else {
-    task = _queue->pull();
+    if (_queue->semaphore.tryWait()) {
+      task = _queue->pull();
+    }
   }
 
   if (!task) {
```

Comment:
No sure if I like removing all comments here. Still would prefer to have sth saying that we first try to pull from the node-local queue and steal from somewhere else if it is empty.